File: plugins/fr.cea.nabla.javalib/src/fr/cea/nabla/javalib/mesh/CartesianMesh2DFactory.java
/*******************************************************************************
 * Copyright (c) 2021 CEA
 * This program and the accompanying materials are made available under the 
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * SPDX-License-Identifier: EPL-2.0
 * Contributors: see AUTHORS file
 *******************************************************************************/
package fr.cea.nabla.javalib.mesh;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

/* Numbering nodes and cells			Numbering faces
 * 
 *  15---16---17---18---19          |-27-|-28-|-29-|-30-|
 *   | 8  | 9  | 10 | 11 |         19   21   23   25   26
 *  10---11---12---13---14          |-18-|-20-|-22-|-24-|
 *   | 4  | 5  | 6  | 7  |         10   12   14   16   17
 *   5----6----7----8----9          |--9-|-11-|-13-|-15-|
 *   | 0  | 1  | 2  | 3  |          1    3    5    7    8
 *   0----1----2----3----4          |-0--|-2--|-4--|-6--|
 */
public class CartesianMesh2DFactory
{
	private int nbXQuads = -1;
	private int nbYQuads = -1;
	private double xSize = -1;
	private double ySize = -1;

	/** Constructor for Json creation */
	public CartesianMesh2DFactory() {}

	public CartesianMesh2DFactory(int nbXQuads, int nbYQuads, double xSize, double ySize)
	{
		this.nbXQuads = nbXQuads;
		this.nbYQuads = nbYQuads;
		this.xSize = xSize;
		this.ySize = ySize;
	}

	public void jsonInit(final String jsonContent)
	{
		final JsonObject o = JsonParser.parseString(jsonContent).getAsJsonObject();
		assert(o.has("nbXQuads"));
		final JsonElement valueof_nbXQuads = o.get("nbXQuads");
		assert(valueof_nbXQuads.isJsonPrimitive());
		nbXQuads = valueof_nbXQuads.getAsJsonPrimitive().getAsInt();
		assert(o.has("nbYQuads"));
		final JsonElement valueof_nbYQuads = o.get("nbYQuads");
		assert(valueof_nbYQuads.isJsonPrimitive());
		nbYQuads = valueof_nbYQuads.getAsJsonPrimitive().getAsInt();
		assert(o.has("xSize"));
		final JsonElement valueof_xSize = o.get("xSize");
		assert(valueof_xSize.isJsonPrimitive());
		xSize = valueof_xSize.getAsJsonPrimitive().getAsDouble();
		assert(o.has("ySize"));
		final JsonElement valueof_ySize = o.get("ySize");
		assert(valueof_ySize.isJsonPrimitive());
		ySize = valueof_ySize.getAsJsonPrimitive().getAsDouble();
	}

	public CartesianMesh2D create()
	{
		if (nbXQuads == -1 || nbYQuads == -1 || xSize == -1 || ySize == -1)
			throw new RuntimeException("MeshFactory attributes uninitialized");

		double[][] nodes = new double[(nbXQuads + 1) * (nbYQuads + 1)][2];
		Quad[] quads = new Quad[nbXQuads * nbYQuads];
		Edge[] edges = new Edge[2 * quads.length + nbXQuads + nbYQuads];

		int[] outerNodesIds = new int[2 * (nbXQuads + nbYQuads)];
		int[] innerNodeIds = new int[nodes.length - outerNodesIds.length];
		int[] topNodeIds = new int[nbXQuads + 1];
		int[] bottomNodeIds = new int[nbXQuads + 1];
		int[] leftNodeIds = new int[nbYQuads + 1];
		int[] rightNodeIds = new int[nbYQuads + 1];

		int[] innerCellIds = new int[(nbXQuads - 2)*(nbYQuads - 2)];
		int[] outerCellIds = new int[2 * nbXQuads + 2 * (nbYQuads - 2)];

		int nodeId = 0;
		int innerNodeId = 0;
		int topNodeId = 0;
		int bottomNodeId = 0;
		int leftNodeId = 0;
		int rightNodeId = 0;

		// node creation
		for (int j = 0; j <=nbYQuads; j++)
			for (int i = 0; i<=nbXQuads; i++)
			{
				nodes[nodeId][0] = xSize * i;
				nodes[nodeId][1] = ySize * j;
				if (i!=0 && j!=0 && i!=nbXQuads && j!=nbYQuads)
					innerNodeIds[innerNodeId++] = nodeId;
				else
				{
					if (j==0) bottomNodeIds[bottomNodeId++] =nodeId;
					if (j==nbYQuads) topNodeIds[topNodeId++] = nodeId;
					if (i==0) leftNodeIds[leftNodeId++] = nodeId;
					if (i==nbXQuads) rightNodeIds[rightNodeId++] = nodeId;
				}
				nodeId++;
		}

		// edge creation
		int nbXNodes = nbXQuads+1;
		int edgeId = 0;
		for (int i = 0; i<nodes.length; i++)
		{
			int rightNodeIndex = i+1;
			if (rightNodeIndex%nbXNodes!=0) edges[edgeId++] = new Edge(i, rightNodeIndex);
			int belowNodeIndex = i + nbXNodes;
			if (belowNodeIndex<nodes.length) edges[edgeId++] = new Edge(i, belowNodeIndex);
		}

		// quad creation
		int quadId = 0;
		int innerCellId = 0;
		int outerCellId = 0;
		for (int j = 0; j < nbYQuads; j++)
			for (int i = 0; i < nbXQuads; i++)
			{
				if( (i != 0) && (i != nbXQuads - 1) && (j != 0) && (j!= nbYQuads - 1) )
					innerCellIds[innerCellId++] = quadId;
				else
					outerCellIds[outerCellId++] = quadId;

				int upperLeftNodeIndex = (j*nbXNodes)+i;
				int lowerLeftNodeIndex = upperLeftNodeIndex + nbXNodes;
				quads[quadId++] = new Quad(upperLeftNodeIndex, upperLeftNodeIndex+1, lowerLeftNodeIndex+1, lowerLeftNodeIndex);
			}

		MeshGeometry meshGeometry = new MeshGeometry(nodes, edges, quads);
		return new CartesianMesh2D(meshGeometry, innerNodeIds, topNodeIds, bottomNodeIds, leftNodeIds, rightNodeIds, innerCellIds, outerCellIds);
	}
}

File: plugins/fr.cea.nabla.cpplib/src/nablalib/mesh/CartesianMesh2DFactory.cc
/*******************************************************************************
 * Copyright (c) 2020 CEA
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * SPDX-License-Identifier: EPL-2.0
 * Contributors: see AUTHORS file
 *******************************************************************************/
#include <rapidjson/document.h>
#include "nablalib/mesh/CartesianMesh2D.h"
#include "nablalib/mesh/CartesianMesh2DFactory.h"

namespace nablalib::mesh
{
  
  /*
   *  15---16---17---18---19          |-27-|-28-|-29-|-30-|
   *   | 8  | 9  | 10 | 11 |         19   21   23   25   26
   *  10---11---12---13---14          |-18-|-20-|-22-|-24-|
   *   | 4  | 5  | 6  | 7  |         10   12   14   16   17
   *   5----6----7----8----9          |--9-|-11-|-13-|-15-|
   *   | 0  | 1  | 2  | 3  |          1    3    5    7    8
   *   0----1----2----3----4          |-0--|-2--|-4--|-6--|
   */

void
CartesianMesh2DFactory::jsonInit(const char* jsonContent)
{
	rapidjson::Document document;
	assert(!document.Parse(jsonContent).HasParseError());
	assert(document.IsObject());
	const rapidjson::Value::Object& o = document.GetObject();

	assert(o.HasMember("nbXQuads"));
	const rapidjson::Value& valueof_nbXQuads = o["nbXQuads"];
	assert(valueof_nbXQuads.IsInt());
	nbXQuads = valueof_nbXQuads.GetInt();

	assert(o.HasMember("nbYQuads"));
	const rapidjson::Value& valueof_nbYQuads = o["nbYQuads"];
	assert(valueof_nbYQuads.IsInt());
	nbYQuads = valueof_nbYQuads.GetInt();

	assert(o.HasMember("xSize"));
	const rapidjson::Value& valueof_xSize = o["xSize"];
	assert(valueof_xSize.IsDouble());
	xSize = valueof_xSize.GetDouble();

	assert(o.HasMember("ySize"));
	const rapidjson::Value& valueof_ySize = o["ySize"];
	assert(valueof_ySize.IsDouble());
	ySize = valueof_ySize.GetDouble();
}

CartesianMesh2D*
CartesianMesh2DFactory::create()
{
	vector<RealArray1D<2>> nodes_((nbXQuads + 1) * (nbYQuads + 1));
	vector<Quad> quads_(nbXQuads * nbYQuads);
	vector<Edge> edges_(2 * quads_.size() + nbXQuads + nbYQuads);

	vector<Id> outer_node_ids_(2 * (nbXQuads + nbYQuads));
	vector<Id> inner_node_ids_(nodes_.size() - outer_node_ids_.size());
	vector<Id> top_node_ids_(nbXQuads + 1);
	vector<Id> bottom_node_ids_(nbXQuads + 1);
	vector<Id> left_node_ids_(nbYQuads + 1);
	vector<Id> right_node_ids_(nbYQuads + 1);

  size_t nb_inner_cells(nbXQuads>1&&nbYQuads>1?(nbXQuads-2)*(nbYQuads-2):0);  // 0 for mesh with only 1 cell
  size_t nb_outer_cells(nbXQuads * nbYQuads - nb_inner_cells);
  vector<Id> inner_cells_ids_(nb_inner_cells);
  vector<Id> outer_cells_ids_(nb_outer_cells);

	// node creation
	Id node_id_(0);
	Id inner_node_id_(0);
	Id top_node_id_(0);
	Id bottom_node_id_(0);
	Id left_node_id_(0);
	Id right_node_id_(0);
	Id top_left_node_id_(0);
	Id top_right_node_id_(0);
	Id bottom_left_node_id_(0);
	Id bottom_right_node_id_(0);
  
	for(size_t j(0); j <= nbYQuads; ++j) {
		for(size_t i(0); i <= nbXQuads; ++i) {
			nodes_[node_id_] = RealArray1D<2>{xSize * i, ySize * j};
			if (i!=0 && j!=0 && i!=nbXQuads && j!=nbYQuads)
				inner_node_ids_[inner_node_id_++] = node_id_;
			else
			{
				if (j==0) bottom_node_ids_[bottom_node_id_++] = node_id_;
				if (j==nbYQuads) top_node_ids_[top_node_id_++] = node_id_;
				if (i==0) left_node_ids_[left_node_id_++] = node_id_;
				if (i==nbXQuads) right_node_ids_[right_node_id_++] = node_id_;
				if (i==0 && j==0) bottom_left_node_id_ = node_id_;
				if (i==nbXQuads && j==0) bottom_right_node_id_ = node_id_;
				if (i==0 && j==nbYQuads) top_left_node_id_ = node_id_;
				if (i==nbXQuads && j==nbYQuads) top_right_node_id_ = node_id_;
			}
			++node_id_;
		}
	}

	// edge creation
	const size_t nb_x_nodes_(nbXQuads + 1);
	Id edge_id_(0);
	for(size_t i(0); i < nodes_.size(); ++i)	{
		const size_t right_node_index_(i + 1);
		if (right_node_index_ % nb_x_nodes_ != 0)
		  edges_[edge_id_++] = move(Edge(static_cast<Id>(i), right_node_index_));
		const size_t above_node_index_(i + nb_x_nodes_);
		if (above_node_index_ < nodes_.size())
		  edges_[edge_id_++] = Edge(static_cast<Id>(i), above_node_index_);
	}

	// quad creation
	Id quad_id_(0);
	Id inner_id_(0);
	Id outer_id_(0);
	for(size_t j(0); j < nbYQuads; ++j) {
		for(size_t i(0); i < nbXQuads; ++i) {
			if( (i != 0) && (i != nbXQuads - 1) && (j != 0) && (j!= nbYQuads - 1) )
			{
				inner_cells_ids_[inner_id_++] = quad_id_;
			}
			else
			{
				outer_cells_ids_[outer_id_++] = quad_id_;
			}
			const size_t upper_left_node_index_((j * static_cast<size_t>(nb_x_nodes_)) + i);
			const size_t lower_left_node_index_(upper_left_node_index_ + static_cast<size_t>(nb_x_nodes_));
			quads_[quad_id_++] = move(Quad(upper_left_node_index_, upper_left_node_index_ + 1,
                                     lower_left_node_index_ + 1, lower_left_node_index_));
		}
	}

	auto mesh_geometry = new MeshGeometry<2>(nodes_, edges_, quads_);
	return new CartesianMesh2D(mesh_geometry, inner_node_ids_, 
		                         top_node_ids_, bottom_node_ids_, 
		                         left_node_ids_, right_node_ids_,
		                         top_left_node_id_, top_right_node_id_,
		                         bottom_left_node_id_, bottom_right_node_id_,
					 inner_cells_ids_, outer_cells_ids_);
}

}
